src/consumer.py

```diff
@@ -2,51 +2,77 @@
 import threading
 import time
 import logging
+import hashlib
 
 from zk_helpers import get_zookeeper_hosts, makeHostsString
 from pubsubClient import PubSubClient
 from chord_node import *
 
-
 class Consumer():
 
     def __init__(self, topics, psclient: PubSubClient) -> None:
-        self.topics = topics     # List of topic names it will publish to
-        self.psclient = psclient # PubSubClient
+        self.topics = topics        # List of topic names it will publish to
+        self.psclient = psclient    # PubSubClient
 
-    def run(self, timeout: int):
+        self.messages = {}
+        self.messages_consumed = {} # List of messages published by this publisher
+        self.messages_digest = {}   # Hash digest of the messages published on this topic
 
+    def run(self, timeout: int):
         # spawn a daemon thread for each topic provided
+        threads = []
         for topic in self.topics:
-            # Daemon threads will be terminated when the main thread terminates
-            topic_thread = threading.Thread(target=self.consume_events, args=(topic,), daemon=True)
+            self.messages[topic] = []
+            self.messages_consumed[topic] = 0
+            self.messages_digest[topic] = hashlib.sha256()
+            topic_thread = threading.Thread(target=self.consume_events, args=(topic, timeout), daemon=True)
             topic_thread.start()
-
-        time.sleep(timeout)
+            threads.append(topic_thread)
+        [t.join() for t in threads]
```

Comment:
Recently learned the difference between daemon and regular threads, so I thought it was okay to assume they'd be killed. I like this better => passing the timeout down to the thread and having a proper join